charms/layers/jupyter-notebook/reactive/jupyter_notebook.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+# Copyright (C) 2016  Ghent University
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# pylint: disable=c0111,c0103,c0301
+import subprocess
+
+from charmhelpers.core import templating, hookenv, host
+from charmhelpers.core.hookenv import charm_dir, open_port, status_set
+from charms.reactive import hook, when, when_not, set_state
+
+import charms.apt #pylint: disable=e0611,e0401
+
+
+@hook('upgrade-charm')
+def upgrade_charm():
+    hookenv.log("Upgrading Notebook Charm")
+    try:
+        subprocess.check_call(['pip','install','-U','jupyter'])
+    except subprocess.CalledProcessError as exception:
+        hookenv.log(exception.output)
+    
+@when_not('apt.installed.python-pip', 'jupyter-notebook.installed')
+def pre_install():
+    hookenv.log("Pre-install")
+    charms.apt.queue_install(['python-pip','python3-pip'])#pylint: disable=e1101
+
+@when('apt.installed.python-pip')
+@when_not('jupyter-notebook.installed')
+def install_jupyter_notebook():
+    hookenv.log("Install Jupyter-notebook")
+    subprocess.check_call(['pip3','install','--upgrade','pip'])
+    subprocess.check_call(['pip3','install','jupyter'])
+    set_state('jupyter-notebook.installed')
+    
+@when('jupyter-notebook.installed')
+def configure_jupyter_notebook():
+    conf = hookenv.config()
+    user = conf['user']
+    home_dir = '/home/%s' % user
+    jupyter_dir = '%s/.jupyter' % home_dir
+    host.mkdir(jupyter_dir)
+    hookenv.log('Configuring jupyter-notebook upstart')
+    render_api_upstart_template()
+    hookenv.log('Generating jupyter notebook config')
+    render_default_config_template(jupyter_dir)
+    host.service_restart('jupyter')
```

Comment:
It's always best to check if the restart succeeded. I think host.service_start() returns a bool indicating if the restart succeeded.

General rule: Charms shouldn't be allowed to fail silently. Two ways of failing loudly:
- Set a blocked state if the fail is due to bad configuration; so the sysadmin can see the message, and change the config.
- Crash the Charm if the error is unknown or can't be fixed by changing the charm config.